pkg/kv/kvserver/rditer/replica_data_iter.go

```diff
@@ -333,95 +346,124 @@ func (ri *ReplicaMVCCDataIterator) UnsafeValue() []byte {
 	return ri.it.UnsafeValue()
 }
 
-// NewReplicaEngineDataIterator creates a ReplicaEngineDataIterator for the given replica.
+// NewReplicaEngineDataIterator creates a ReplicaEngineDataIterator for the
+// given replica.
 func NewReplicaEngineDataIterator(
-	d *roachpb.RangeDescriptor, reader storage.Reader, replicatedOnly bool,
+	desc *roachpb.RangeDescriptor, reader storage.Reader, replicatedOnly bool,
 ) *ReplicaEngineDataIterator {
-	it := reader.NewEngineIterator(storage.IterOptions{UpperBound: d.EndKey.AsRawKey()})
+	if !reader.ConsistentIterators() {
+		panic("ReplicaEngineDataIterator requires consistent iterators")
+	}
 
-	rangeFunc := MakeAllKeyRanges
+	var ranges []KeyRange
 	if replicatedOnly {
-		rangeFunc = MakeReplicatedKeyRanges
+		ranges = MakeReplicatedKeyRanges(desc)
+	} else {
+		ranges = MakeAllKeyRanges(desc)
 	}
-	ri := &ReplicaEngineDataIterator{
-		ranges: rangeFunc(d),
-		it:     it,
+
+	return &ReplicaEngineDataIterator{
+		reader: reader,
+		ranges: ranges,
 	}
-	ri.seekStart()
-	return ri
 }
 
-// seekStart seeks the iterator to the start of its data range.
-func (ri *ReplicaEngineDataIterator) seekStart() {
-	ri.curIndex = 0
-	ri.valid, ri.err = ri.it.SeekEngineKeyGE(storage.EngineKey{Key: ri.ranges[ri.curIndex].Start})
-	ri.advance()
+// nextIter creates an iterator for the next non-empty key range/type and seeks
+// it, closing the existing iterator if any. Returns false if all key ranges and
+// key types have been exhausted.
+//
+// TODO(erikgrinaker): Rather than creating a new iterator for each key range,
+// we could expose an API to reconfigure the iterator with new bounds. However,
+// the caller could also use e.g. a pebbleReadOnly which reuses the iterator
+// internally. This should be benchmarked.
+func (ri *ReplicaEngineDataIterator) nextIter() (bool, error) {
+	for {
+		if ri.it == nil {
+			ri.curIndex = 0
+			ri.curKeyType = storage.IterKeyTypePointsOnly
+		} else if ri.curKeyType == storage.IterKeyTypePointsOnly {
+			ri.curKeyType = storage.IterKeyTypeRangesOnly
+		} else if ri.curIndex+1 < len(ri.ranges) {
+			ri.curIndex++
+			ri.curKeyType = storage.IterKeyTypePointsOnly
+		} else {
+			break
```

Comment:
I think it's useful to have access to the current configuration of the current iterator. We use it in `HasPointAndRange`, for example. Otherwise we'd have to have logic to backtrack.

Reply:
We can have a switch at the beginning of the loop and start with -1 and storage.IterKeyTypeRangesOnly to address that. 

What I think is that there's a reason why terminating loops in the middle except for obvious cases like errors or one liner bounds checks is hurting readability same way as breaking to labels and the like.